Game board

KingPhilip14 opened the pull request:
I'm not sure if the documentation is too much or simply unclear. Please let me know so I can rewrite it. I'm also not too sure if the occupied_by loop I added to the avatars() and stations() methods is correct. If all is well, the game_board file is done.

KingPhilip14 committed:
Changed generation system to game_board.py

Changed the system of how maps are generated.  A docstring providing how it works will be added later

KingPhilip14 committed:
Merge branch 'master' into game_board

KingPhilip14 committed:
Updated __help method

Renamed __help to __help_populate() and added more functionality to it for checking for occupied_by.

KingPhilip14 committed:
Merge branch 'master' into game_board

KingPhilip14 committed:
Merge branch 'master' into game_board

KingPhilip14 committed:
Updated game_board.py

This isn't done yet, but changes were made to include Vector in some areas.

KingPhilip14 committed:
Updated game_board

Added detailed documentation to explain the new system that was implemented for generation. Also added a method to go through the occupied_by chain in avatars() and stations().

JeanVArdent reviewed (changes requested):
Missing properties

KingPhilip14 committed:
Updated game_board

Added properties for the field variables.

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+
+    @map_size.setter
+    def map_size(self, map_size: Vector):
+        if map_size is None or isinstance(map_size, Vector):
```
This is not going to work as intended, setting the property does not return. Use De Morgan's law to reverse the conditional and raise an exception if that condition is met

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+
+    @locations.setter
+    def locations(self, locations: dict[[Vector]:[GameObject]]):
+        if locations is None or isinstance(locations, dict):
```
Same here

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+
+    @walled.setter
+    def walled(self, walled: bool):
+        if walled is None or isinstance(walled, bool):
```
Same here

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+            if len(k) != len(v) or (len(k) == 0 or len(v) == 0):  # Key-Value lengths must be > 0 and equal
+                raise ValueError("A key-value pair from game_board.locations has mismatching lengths. "
+                                 "They must be the same length, regardless of size.")
+            j = random.choices(k, k=len(k))
```
Give this an inline comment. Someone is bound to be confused by this at some point

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+                    self.__loop_occupied_by(col, to_return)
+        return to_return
+
+    def __loop_occupied_by(self, col: GameObject, to_return: list[GameObject]) -> None:
```
I like the concept behind this method. One, what does col mean? It's a bit of a weird parameter name. Two, this function does not test for a specific type of object using isinstance so you are returning a list of all occupiables, not the specific game object you are looking for.

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+        to_return = list()
+        for row in self.game_map:
+            for col in row:
+                if isinstance(col.occupied_by, Avatar):
```
You'd want to do your occupied by loop before checking for avatars at the tail

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+    def from_json(self, data) -> Self:
+        super().from_json(data)
+        temp = data["game_map"]
+        self.game_map = list([list(map(lambda tile: Tile().from_json(tile), y)) for y in temp])
```
I don't think this second list comprehension to list cast is necessary, whoops

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+                 locations: dict[[Vector]:[GameObject]] = None, walled: bool = False):
+
+        random.seed(seed)
+        super().__init__()
```
This should always be the first line of a constructor

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+                        self.game_map[y][x].occupied_by = Wall()
+                self.game_map[0][x].occupied_by = Wall()
+                self.game_map[self.map_size.y - 1][x].occupied_by = Wall()
+
```
You don't call populate_map in the constructor. Is this intentional?

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+            self.__help_populate(j, v)
+
+    def __help_populate(self, j: Vector, v: list[GameObject]):
+        for i in v:
```
For i, game_object in enumerate (v):... game_object would replace i and i would represent the index to be used on j to get the vectors to be used as j is a list of vectors not a vector

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+
+            temp = self.game_map[j.y][j.x]
+
+            while temp.occupied_by is not None:
```
This loop was actually broken. Please see the fixed loop on Gunnar's PR

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+
+class GameBoard(GameObject):
+    """
+    Notes for creating the GameBoard:
```
This looks great!!

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+        x       x
+        x x x x x   y = 6
+    """
+    def __init__(self, seed: int = None, map_size: Vector = Vector(),
```
Use int|None type hinting for instances where we allow None

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+        x x x x x   y = 6
+    """
+    def __init__(self, seed: int = None, map_size: Vector = Vector(),
+                 locations: dict[[Vector]:[GameObject]] = None, walled: bool = False):
```
Same here, dict[list[Vector]:list[GameObject]]|None would be the type hint

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+                to_return.append(col.occupied_by)
+
+
+    def to_json(self) -> dict:
```
Forgot to type hint in this and the from JSON functions

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,221 @@
+
+    def to_json(self) -> dict:
+        data = super().to_json()
+        temp = list([list(map(lambda tile: tile.to_json(), y)) for y in self.game_map])
```
List comprehension to list cast probably not needed here

JeanVArdent reviewed (changes requested):
Finally got time to really look it over. As a tip, remember to type hint at every instance you would have to declare a variable for the first time. All in all this is a really complicated class and you are off to a great start.

KingPhilip14 committed:
Finalized game_board

Consolidated avatars() and stations() into   get_objects(). Made generate map() a stand-alone method as well and use it in generate_game.py.

JeanVArdent commented on game/utils/generate_game.py:
```diff
@@ -2,15 +2,19 @@
     print('Generating game map...')
 
-    data = dict()
+    temp = GameBoard(seed)
```
Type hint

JeanVArdent commented on game/utils/generate_game.py:
```diff
@@ -2,15 +2,19 @@
-    data = dict()
+    temp = GameBoard(seed)
+    temp.generate_map()
+    data = temp.to_json()
```
Type hint

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,228 @@
+    @seed.setter
+    def seed(self, seed: int | None):
+        if seed is not None and not isinstance(seed, int):
+            raise ValueError("Seed must be an integer.")
```
Or None

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,228 @@
+
+    @map_size.setter
+    def map_size(self, map_size: Vector):
+        if map_size is None or isinstance(map_size, Vector):
```
Not going to work as intended

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,228 @@
+    def locations(self, locations: dict[[Vector]:[GameObject]]):
+        if locations is None or isinstance(locations, dict):
+            self.__locations = locations
+        raise ValueError("Locations must be a dict. The key must be a list of Vector Objects, and the "
```
Or None

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,228 @@
+
+    @walled.setter
+    def walled(self, walled: bool):
+        if walled is None or isinstance(walled, bool):
```
Do not permit None

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,228 @@
+
+        for row in self.game_map:
+            for object_in_row in row:
+                temp: GameObject | Tile = object_in_row
```
Tile is a GameObject so | Tile is not needed here

JeanVArdent reviewed (changes requested):
You forgot a few type hinting things and to switch the rest of your properties around, but other than that the rest of the logic looks fine to me!

KingPhilip14 committed:
Update game_board.py

Started making the suggested changes to game_board. It's not finished yet

KingPhilip14 committed:
Conflict resolutions

I hopefully fixed all the issues wrong with game_board. The properties and type hinting should all be good.

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,230 @@
+
+    @locations.setter
+    def locations(self, locations: dict[[Vector]:[GameObject]] | None):
+        if locations is not None or not isinstance(locations, dict):
```
Check if the len of each key is equal to the len of each value when setting as well

JeanVArdent reviewed (changes requested):
Just one small thing to consider with the locations setter. See my comment for more details.

KingPhilip14 committed:
Finalized game_board

Since lists can't be used as a key in a dict, it was changed to be a tuple. Both documentation and methods were adjusted for this.

JeanVArdent commented on game/common/game_board.py:
```diff
@@ -0,0 +1,249 @@
+
+        return to_return
+
+    @staticmethod
```
I wouldn't make this a static method

JeanVArdent reviewed (changes requested):
Just remove the static method decorator and it looks good to me

KingPhilip14 committed:
Removed Static Method

Removed the static tag from a static helper method.

JeanVArdent reviewed (approved):
Lgtm

ESatrom reviewed (approved):
look good
